citus 5.0.1 (new formula)

jasonmp85 opened the pull request:
Today Citus Data announced they "unforked" their PostgreSQL-based distributed database and open-sourced the whole project. This means Citus, previously a private fork of the PostgreSQL codebase, can now be added to an existing PostgreSQL install using the standard PostgreSQL CREATE EXTENSION command.

This is analogous to e.g. PostGIS, whose formula was used as an example when writing the formula for Citus. Other PostgreSQL-related formula provided good examples of a test block that actually starts PostgreSQL and verifies installation of the extension.

## Code Review Tasks

  - [x] Set `pg_config` using environment variable
  - [x] Move `make` and `configure` args inline to the relevant `system` calls
  - [x] Bump to latest Citus version (v5.0.1)
  - [x] Investigate installing directly to Cellar
  - [x] Move to new Homebrew repo and close this issue

dunn commented on Library/Formula/citus.rb:
```diff
@@ -0,0 +1,55 @@
+  depends_on "postgresql"
+
+  def install
+    config_args = %W[--prefix=#{prefix} PG_CONFIG=#{Formula["postgresql"].opt_bin}/pg_config]
```
For `pg_config` can it be set as an environment variable instead? Like 
```ruby
ENV["PG_CONFIG"] = Formula["postgresql"].opt_bin}/"pg_config"

jasonmp85 replied:
I can see if that would work.

jasonmp85 replied:
Done.

dunn commented on Library/Formula/citus.rb:
```diff
@@ -0,0 +1,55 @@
+    # workaround for https://github.com/Homebrew/homebrew/issues/49948
+    make_args = ["libpq=-L#{Formula["postgresql"].opt_lib} -lpq"]
+
+    system "./configure", *config_args
```
Not sure if it's worth setting up make and configure `args` arrays if they never change.

jasonmp85 replied:
Sure.

jasonmp85 replied:
Done. In fact, the `configure` argument was entirely unnecessary, so I removed it.

dunn commented on Library/Formula/citus.rb:
```diff
@@ -0,0 +1,55 @@
+    system "make", *make_args
+
+    mkdir "stage"
+    system "make", "install", "DESTDIR=#{buildpath}/stage"
```
Why can't we `make install` directly into the Cellar?

jasonmp85 replied:
Was using PostGIS as a reference and it's what it did. It worked, so I didn't really mess with it. I can try installing directly, if desired.

dunn replied:
Yeah, if it works.  Though looking closer at `Dir["stage/**/bin/*"]` below, what directories are captured by the glob?

jasonmp85 replied:
So here's the deal:

Even though we have a `configure` script, `make install` doesn't really honor the `prefix` (which is why I removed it: it was misleading).

Instead, the PostgreSQL extension building support (documented [here](http://www.postgresql.org/docs/9.5/static/extend-pgxs.html)) will call `pg_config` to determine the bin, include, lib, and share directories known to the PostgreSQL installation.

Unfortunately, some of these are top-level (as in `/usr/local/lib`) and most others are buried within the PostgreSQL package's own Cellar (i.e. underneath `/usr/local/Cellar/postgresql/9.5.2/`).

_Even_ with a `DESTDIR` argument, PGXS will recreate this hierarchy underneath the destdir, which is the reason we need that glob. The binaries are actually at `$DESTDIR/usr/local/Cellar/postgresql/9.5.2/bin`.

I experimented with trying to force PGXS to just install the files directly within the package's Cellar, but its integration with `pg_config` is just too tight. As _all_ other PostgreSQL extension Homebrew formulas built using PGXS (PostGIS, pldebugger, pgtap, pgrouting, and pgroonga) have a `mkdir 'stage'` line, I think it would be _more_ confusing to have this formula try to deviate from the norm they establish.

So essentially the answer is "PostgreSQL has an extension building infrastructure that bypasses everything and installs files directly into the directories reported by `pg_config`, so we first install to a `DESTDIR` and then map that hierarchy to something more Homebrew-friendly".

FWIW, the RPM and Debian builds I authored had to do much the same thing: build the extension, install to a `DESTDIR`, grab the files out of there into a structure `yum` or `apt` understand, then create the `rpm` or `deb` package.

apjanke replied:
That makes sense. Especially if it's established practice, we can go with this until someone finds an alternate mechanism for all the PostgreSQL plugin builds.

jasonmp85 commented:
@dunn — I addressed your feedback (checklist up top) and updated the Citus version. The `stage` directory turns out to just be more straightforward than any alternatives, and is used by _all_ other Homebrew-provided, PGXS-based PostgreSQL extension installs. So I've kept it.

Shall I reopen this under the new repo (I will unless you object)?

jasonmp85 commented:
Closing in favor of the new Homebrew/homebrew-core#463.
